synapse/http/matrixfederationclient.py

```diff
@@ -184,20 +184,54 @@ def get_json(self) -> Optional[JsonDict]:
         return self.json
 
 
-class JsonParser(ByteParser[Union[JsonDict, list]]):
+class _BaseJsonParser(ByteParser[T]):
     """A parser that buffers the response and tries to parse it as JSON."""
 
     CONTENT_TYPE = "application/json"
 
-    def __init__(self) -> None:
+    def __init__(self, validator: Optional[Callable[[Any], bool]] = None) -> None:
         self._buffer = StringIO()
         self._binary_wrapper = BinaryIOWrapper(self._buffer)
+        self._validator = validator
 
     def write(self, data: bytes) -> int:
         return self._binary_wrapper.write(data)
 
-    def finish(self) -> Union[JsonDict, list]:
-        return json_decoder.decode(self._buffer.getvalue())
+    def finish(self) -> T:
+        result = json_decoder.decode(self._buffer.getvalue())
+        if self._validator is not None and not self._validator(result):
+            raise ValueError("Unexpected JSON object")
```

Comment:
If JsonDictParser is given a list, it'll complain that it recieved an unexpected JSON object, which seems wrong. Should we include the `result` in the error message too?

Perhaps `f"Malformed JSON value: {result}"`?

Reply:
> If JsonDictParser is given a list, it'll complain that it recieved an unexpected JSON object, which seems wrong. Should we include the `result` in the error message too?

Why is this wrong? A JSON array isn't an object.

Reply:
Do you mean that I abstracted the validator, but not the error message?

Reply:
> > If JsonDictParser is given a list, it'll complain that it recieved an unexpected JSON object, which seems wrong. Should we include the `result` in the error message too?
> 
> Why is this wrong? A JSON array isn't an object.

If I expect a Dict/object and I get a List/array, the error message should say "unexpected array", not "unexpected object". Or better: "Expected object, but got an array". The current behaviour:

```python
>>> from synapse.http.matrixfederationclient import JsonDictParser
>>> p = JsonDictParser()
>>> p.write(b"[]")
2
>>> p.finish()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dmr/workspace/synapse/synapse/http/matrixfederationclient.py", line 203, in finish
    raise ValueError("Unexpected JSON object")
ValueError: Unexpected JSON object
```

> Do you mean that I abstracted the validator, but not the error message?

Not worried about abstracting out the error message, I just want to ensure the error message won't confuse our future selves.

Reply:
```suggestion
            raise ValueError(f"Received incorrect JSON value: {result.__class__.__name__}")
```

Reply:
Works for me (sorry for digging in on this). Looks like a test might be suddenly failing though?